### PR TITLE
Use parameterized time_bucket for chunk exclusion

### PR DESCRIPTION
**Pull request by akuzm**
When time_bucket is compared to constant in WHERE, we also add condition on the underlying time variable (ts_transform_time_bucket_comparison). Unfortunately we only do this for plan-time constants, which prevents chunk exclusion when the interval is given by a query parameter, and a generic plan is used. This commit also tries to apply this optimization after applying the execution-time constants.

This PR also enables startup exclusion based on parameterized filters.

Fixes https://github.com/timescale/timescaledb/issues/6498

Disable-check: force-changelog-file
Disable-check: commit-count

### CODE REVIEW

File: test/expected/plan_expand_hypertable-16.out [hunk 1]
```diff
@@ -1217,6 +1217,157 @@ transformation would be out of range
...
+(3 rows)
+
+:PREFIX EXECUTE P5('60 mins');
+           QUERY PLAN           
+--------------------------------
+ Sort
+   Sort Key: "time"
+   ->  Result
+         One-Time Filter: false
+(4 rows)
```
Comment (akuzm) [inline test/expected/plan_expand_hypertable-16.out:1270]: This query has only immutable operators, so with forced custom plans it uses plan-time exclusion. Same for P8.

**Comment by svenklemm** [comment]
Would be good to also have some tests where not all chunks get excluded

File: src/nodes/chunk_append/exec.c [hunk 2]
```diff
@@ -925,16 +927,42 @@ chunk_append_get_lock_pointer()
...
-		rinfo->clause = (Expr *) estimate_expression_value(root, (Node *) rinfo->clause);
+		if (!equal(rinfo->clause, constified))
+		{
+			/*
+			 * We have constified something, so try applying the time_bucket
+			 * transformations again. This might allow us to exclude chunks
+			 * based on a parameterized time_bucket expression.
+			 */
+			Expr *additional_clause = ts_transform_time_bucket_comparison(constified);
+			if (additional_clause != constified)
```
Comment (erimatnor) [inline src/nodes/chunk_append/exec.c:947]: Is it enough to do a pointer comparison here? It seems that the expression could have been changed within but returning the same pointer.

Above you are using the `equal` function so it seems inconsistent.

File: src/planner/expand_hypertable.c [hunk 3]
```diff
@@ -660,31 +693,30 @@ process_quals(Node *quals, CollectQualCtx *ctx, bool is_outer_join)
...
+			else
+			{
+				/*
+				 * check for time_bucket comparisons
+				 * time_bucket(Const, time_colum) > Const
 				 */
-				if (qual != (Expr *) op)
+				Expr *original = qual;
+				qual = ts_transform_time_bucket_comparison(qual);
+				if (qual != original)
```
Comment (erimatnor) [inline src/planner/expand_hypertable.c:712]: Same concern here about pointer comparison.

File: src/nodes/chunk_append/exec.c (hunk 2 above)
Comment (akuzm) [reply src/nodes/chunk_append/exec.c:947]: Above I have to use `equal` because it `estimate_expression_value` uses expression mutators which always return a deep copy of the object, even if it's not modified. The time bucket transform is much simpler and returns original object if it's not modified.

Comment (akuzm) [reply src/nodes/chunk_append/exec.c]: I made the time bucket transform return NULL when it doesn't apply, to avoid confusion.

File: src/planner/expand_hypertable.c [hunk 4]
```diff
@@ -385,25 +385,59 @@ constify_timestamptz_op_interval(PlannerInfo *root, OpExpr *constraint)
  * Expressions with value on the left side will be switched around
  * when building the expression for RestrictInfo.
  *
- * Caller must ensure that only 2 argument time_bucket versions
```
Comment (konskov) [inline src/planner/expand_hypertable.c:388]: This comment is still relevant, right? Only 2 argument time_bucket versions are acceptable

Comment (akuzm) [reply src/planner/expand_hypertable.c:388]: Not really, the caller doesn't have to ensure this anymore. I moved these check inside this function itself.

File: src/planner/expand_hypertable.c [hunk 5]
```diff
@@ -660,31 +693,34 @@ process_quals(Node *quals, CollectQualCtx *ctx, bool is_outer_join)
...
-				 list_length(castNode(FuncExpr, right)->args) == 2 &&
-				 is_time_bucket_function(right)))
 			{
-				qual = (Expr *) transform_time_bucket_comparison(ctx->root, op);
 				/*
-				 * if we could transform the expression we add it to the list of
-				 * quals so it can be used as an index condition
+				 * check for constraints with TIMESTAMPTZ OP INTERVAL calculations
+				 */
+				qual = &constify_timestamptz_op_interval(ctx->root, op)->xpr;
```
Comment (konskov) [inline src/planner/expand_hypertable.c:702]: why is this changed from `qual = (Expr *) constify_timestamptz_op_interval(ctx->root, op);`

**Review by konskov — APPROVED** [review]

File: src/planner/expand_hypertable.c (hunk 5 above)
Comment (akuzm) [reply src/planner/expand_hypertable.c:702]: I think it's a leftover from back and forth changes, I'll remove it (it's the same cast just written differently).

**Review by svenklemm — APPROVED** [review]
